File: Sharebook/Server/src/Sharebook.Common/Iterators/MultiLevelCursorIterator.cs
namespace Sharebook.Common.Iterators
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Threading.Tasks;

    using Sharebook.Common.Exceptions.Public;
    using Sharebook.Common.Extensions;
    using Sharebook.Common.Iterators.Models;
    using Sharebook.Common.Public.Iterators;

    public class MultiLevelCursorIterator<TParent, TItem> : MultiLevelCursorIterator<TParent, TItem, string>, IMultiLevelCursorIterator<TParent, TItem>
    {
        public MultiLevelCursorIterator(
            Func<Task<ICursorPageResult<TParent, string>>> iterateSubLevel,
            Func<TParent[], Task<IPageProcessingResult<TParent, TItem>>> getChildItemsPageFromParent) : base(iterateSubLevel, getChildItemsPageFromParent)
        {
        }
    }

    public class MultiLevelCursorIterator<TParent, TItem, TCursor> : IMultiLevelCursorIterator<TParent, TItem, TCursor>
    {
        private readonly Func<Task<ICursorPageResult<TParent, TCursor>>> iterateSubLevel;
        private readonly Func<TParent[], Task<IPageProcessingResult<TParent, TItem>>> getChildItemsPageFromParent;

        private HashSet<TParent> itemsLeftToProcess;
        private ICursorPageResult<TParent, TCursor> lastParentPageResult;

        public MultiLevelCursorIterator(
            Func<Task<ICursorPageResult<TParent, TCursor>>> iterateSubLevel,
            Func<TParent[], Task<IPageProcessingResult<TParent, TItem>>> getChildItemsPageFromParent)
        {
            this.iterateSubLevel = iterateSubLevel;
            this.getChildItemsPageFromParent = getChildItemsPageFromParent;
            this.itemsLeftToProcess = new HashSet<TParent>();
        }

        public bool Completed => this.lastParentPageResult != null && this.lastParentPageResult.IsLastPage && this.itemsLeftToProcess.Count == 0;

        public async Task<IMultiLevelCursorIteratorPage<TParent, TItem, TCursor>> NextPageAsync()
        {
            if (Completed)
            {
                throw new TwitterIteratorAlreadyCompletedException();
            }

            if (this.lastParentPageResult == null || this.itemsLeftToProcess.Count == 0)
            {
                this.lastParentPageResult = await this.iterateSubLevel().ConfigureAwait(false);
                this.itemsLeftToProcess = new HashSet<TParent>(this.lastParentPageResult.Items);
            }

            var childItemsPage = await this.getChildItemsPageFromParent(this.itemsLeftToProcess.ToArray()).ConfigureAwait(false);
            var processedParentItems = childItemsPage.AssociatedParentItems;

            processedParentItems.ForEach(item => { this.itemsLeftToProcess.Remove(item); });

            var pageResult = new MultiLevelCursorIteratorPage<TParent, TItem, TCursor>
            {
                Items = childItemsPage.Items,
                AssociatedParentItems = processedParentItems,
                NextCursor = this.lastParentPageResult.NextCursor,
                IsLastPage = Completed
            };

            return pageResult;
        }
    }
}

File: Sharebook/Client/src/app/core/Core/Iterators/MultiLevelCursorIterator.ts
import {ICursorPageResult} from "./Models/CursorPageResult";
import {IPageProcessingResult} from "./Models/MultiLevelPageProcessingResult";
import HashSet from "../../../c#-objects/TypeScript.NET-Core/packages/Core/source/Collections/HashSet";
import {TwitterIteratorAlreadyCompletedException} from "../../Public/Exceptions/TwitterIteratorAlreadyCompletedException";
import {IMultiLevelCursorIteratorPage} from "../../Public/Iterators/IMultiLevelCursorIteratorPage";
import {MultiLevelCursorIteratorPage} from "./Models/MultiLevelCursorIteratorPage";
import {IMultiLevelCursorIterator} from "../../Public/Iterators/IMultiLevelCursorIterator";

// TODO: check C# code!!!!
export class MultiLevelCursorIteratorBase<TParent, TItem, TCursor> implements IMultiLevelCursorIterator<TParent, TItem, TCursor> {
  private readonly _iterateSubLevel: () => Promise<ICursorPageResult<TParent, TCursor>>;
  private readonly _getChildItemsPageFromParent: (parents: TParent[]) => Promise<IPageProcessingResult<TParent, TItem>>;

  private _itemsLeftToProcess: HashSet<TParent>;
  private _lastParentPageResult: ICursorPageResult<TParent, TCursor>;

  constructor(iterateSubLevel: () => Promise<ICursorPageResult<TParent, TCursor>>,
              getChildItemsPageFromParent: (TParent: TParent[]) => Promise<IPageProcessingResult<TParent, TItem>>) {
    this._iterateSubLevel = iterateSubLevel;
    this._getChildItemsPageFromParent = getChildItemsPageFromParent;
    this._itemsLeftToProcess = new HashSet<TParent>();
  }

  get completed(): boolean {
    return this._lastParentPageResult != null && this._lastParentPageResult.isLastPage && this._itemsLeftToProcess.count === 0;
  }

  public async nextPageAsync(): Promise<IMultiLevelCursorIteratorPage<TParent, TItem, TCursor>> {
    if (this.completed) {
      throw new TwitterIteratorAlreadyCompletedException();
    }

    if (this._lastParentPageResult == null || this._itemsLeftToProcess.count === 0) {
      this._lastParentPageResult = await this._iterateSubLevel(); // .ConfigureAwait(false);
      this._itemsLeftToProcess = new HashSet<TParent>(this._lastParentPageResult.items);
    }

    let childItemsPage = await this._getChildItemsPageFromParent(this._itemsLeftToProcess.toArray()); // .ConfigureAwait(false);
    let processedParentItems = childItemsPage.associatedParentItems;
    processedParentItems.forEach(item => { this._itemsLeftToProcess.remove(item); }); // TODO: might bug

    let pageResult = new MultiLevelCursorIteratorPage<TParent, TItem, TCursor>();
    pageResult.items = childItemsPage.items;
    pageResult.associatedParentItems = processedParentItems;
    pageResult.nextCursor = this._lastParentPageResult.nextCursor;
    pageResult.isLastPage = this.completed;

    return pageResult;
  }
}

export class MultiLevelCursorIterator<TParent, TItem> extends MultiLevelCursorIteratorBase<TParent, TItem, string> implements IMultiLevelCursorIterator<TParent, TItem> {
  constructor(iterateSubLevel: () => Promise<ICursorPageResult<TParent, string>>,
              getChildItemsPageFromParent: (tParent: TParent[]) => Promise<IPageProcessingResult<TParent, TItem>>) {
    super(iterateSubLevel, getChildItemsPageFromParent);
  }
}

